src/main/java/uk/gov/ch/oauth/Oauth2.java

```diff
@@ -142,29 +136,29 @@ private String getSessionNonce() {
         return oauth2Nonce;
     }
 
-    public UserProfileResponse getUserProfile(final String code, final Session chSession) {
-        LOGGER.debug("Requesting User Profile");
-
-        final OAuthToken oauthToken = getOAuthToken(code);
-        final WebClient webClient = WebClient.create();
-        final URI profileUrl = URI.create(identityProvider.getProfileUrl());
-
-        final UserProfileResponse userProfile =
-                getUserProfileResponse(oauthToken, webClient, profileUrl);
+    private UserProfileResponse fetchUserProfile(final String code) {
+        final OAuthToken oauthToken = requestOAuthToken(code);
 
+        UserProfileResponse userProfile = requestUserProfile(oauthToken);
+        if ((userProfile.getId() == null) || userProfile.getId().isEmpty()) {
+            return null;
+        }
         final Map<String, Object> signInData = oauthToken.saveAccessToken();
         userProfile.addUserProfileToMap(signInData);
         signInData.put(SessionKeys.SIGNED_IN.getKey(), 1);
         final String signInInfoKey = SessionKeys.SIGN_IN_INFO.getKey();
-        final Map<String, Object> sData = chSession.getData();
+        final Map<String, Object> sData = sessionFactory.getSessionDataFromContext();
```

Comment:
No need to get the session from context again when we already have the session in scope here, please undo.

Reply:
session was removed from the method signature, so its no longer in scope.

Reply:
True, however we are getting the same session when verifying the nonce and then extracting the user profile. Maybe look to set it once.

Reply:
Doing it once within a method would make sense, however we are reducing the coupling of the methods by doing it this way.